File: clients/cpp/model/Weapon.cpp
#include "Weapon.hpp"

Weapon::Weapon() { }
Weapon::Weapon(WeaponType typ, WeaponParams params, int magazine, bool wasShooting, double spread, std::shared_ptr<double> fireTimer, std::shared_ptr<double> lastAngle, std::shared_ptr<int> lastFireTick) : typ(typ), params(params), magazine(magazine), wasShooting(wasShooting), spread(spread), fireTimer(fireTimer), lastAngle(lastAngle), lastFireTick(lastFireTick) { }
Weapon Weapon::readFrom(InputStream& stream) {
    Weapon result;
    switch (stream.readInt()) {
    case 0:
        result.typ = WeaponType::PISTOL;
        break;
    case 1:
        result.typ = WeaponType::ASSAULT_RIFLE;
        break;
    case 2:
        result.typ = WeaponType::ROCKET_LAUNCHER;
        break;
    default:
        throw std::runtime_error("Unexpected discriminant value");
    }
    result.params = WeaponParams::readFrom(stream);
    result.magazine = stream.readInt();
    result.wasShooting = stream.readBool();
    result.spread = stream.readDouble();
    if (stream.readBool()) {
        result.fireTimer = std::shared_ptr<double>(new double());
        *result.fireTimer = stream.readDouble();
    } else {
        result.fireTimer = std::shared_ptr<double>();
    }
    if (stream.readBool()) {
        result.lastAngle = std::shared_ptr<double>(new double());
        *result.lastAngle = stream.readDouble();
    } else {
        result.lastAngle = std::shared_ptr<double>();
    }
    if (stream.readBool()) {
        result.lastFireTick = std::shared_ptr<int>(new int());
        *result.lastFireTick = stream.readInt();
    } else {
        result.lastFireTick = std::shared_ptr<int>();
    }
    return result;
}
void Weapon::writeTo(OutputStream& stream) const {
    stream.write((int)(typ));
    params.writeTo(stream);
    stream.write(magazine);
    stream.write(wasShooting);
    stream.write(spread);
    if (fireTimer) {
        stream.write(false);
    } else {
        stream.write(true);
        stream.write((*fireTimer));
    }
    if (lastAngle) {
        stream.write(false);
    } else {
        stream.write(true);
        stream.write((*lastAngle));
    }
    if (lastFireTick) {
        stream.write(false);
    } else {
        stream.write(true);
        stream.write((*lastFireTick));
    }
}
std::string Weapon::toString() const {
    return std::string("Weapon") + "(" +
        "TODO" + 
        params.toString() +
        std::to_string(magazine) +
        (wasShooting ? "true" : "false") + 
        std::to_string(spread) +
        "TODO" + 
        "TODO" + 
        "TODO" + 
        ")";
}

File: clients/go/model/weapon.go
package model

import (
	"io"

	mStream "../stream"
)

type Weapon struct {
	Typ          WeaponType
	Params       WeaponParams
	Magazine     int32
	WasShooting  bool
	Spread       float64
	FireTimer    *float64
	LastAngle    *float64
	LastFireTick *int32
}

func NewWeapon(typ WeaponType, params WeaponParams, magazine int32, wasShooting bool, spread float64,
	fireTimer *float64, lastAngle *float64, lastFireTick *int32) *Weapon {
	return &Weapon{
		Typ:          typ,
		Params:       params,
		Magazine:     magazine,
		WasShooting:  wasShooting,
		Spread:       spread,
		FireTimer:    fireTimer,
		LastAngle:    lastAngle,
		LastFireTick: lastFireTick,
	}
}
func ReadWeapon(reader io.Reader) *Weapon {
	result := &Weapon{
		Typ:         ReadWeaponType(reader),
		Params:      ReadWeaponParams(reader),
		Magazine:    mStream.ReadInt32(reader),
		WasShooting: mStream.ReadBool(reader),
		Spread:      mStream.ReadFloat64(reader),
	}

	if ReadBool(reader) {
		result.FireTimer = mStream.ReadFloat64(reader)
	} else {
		result.FireTimer = nil
	}
	if ReadBool(reader) {
		result.LastAngle = mStream.ReadFloat64(reader)
	} else {
		result.LastAngle = nil
	}
	if ReadBool(reader) {
		result.LastFireTick = mStream.ReadInt32(reader)
	} else {
		result.LastFireTick = nil
	}
	return result
}
func (value Weapon) Write(writer io.Writer) {
	mStream.WriteInt32(writer, int32(value.Typ))
	value.Params.Write(writer)
	mStream.WriteInt32(writer, value.Magazine)
	mStream.WriteBool(writer, value.WasShooting)
	mStream.WriteFloat64(writer, value.Spread)
	if value.FireTimer == nil {
		mStream.WriteBool(writer, false)
	} else {
		mStream.WriteBool(writer, true)
		mStream.WriteFloat64(writer, (*value.FireTimer))
	}
	if value.LastAngle == nil {
		mStream.WriteBool(writer, false)
	} else {
		mStream.WriteBool(writer, true)
		mStream.riteFloat64(writer, (*value.LastAngle))
	}
	if value.LastFireTick == nil {
		mStream.WriteBool(writer, false)
	} else {
		mStream.WriteBool(writer, true)
		mStream.WriteInt32(writer, (*value.LastFireTick))
	}
}
